MRG: add function to retrieve emptyroom

jasmainak opened the pull request:
PR Description
--------------

closes #287 

Merge checklist
---------------

Maintainer, please confirm the following before merging:

- [x] All comments resolved
- [x] This is not your own PR
- [x] All CIs are happy
- [x] PR title starts with [MRG]
- [x] [whats_new.rst](https://github.com/mne-tools/mne-bids/blob/master/doc/whats_new.rst) is updated
- [x] PR description includes phrase "closes <#issue-number>"
- [x] Commit history does not contain any merge commits

jasmainak commented:
The tests in this PR depend on https://github.com/mne-tools/mne-python/issues/7067

agramfort commented on mne_bids/read.py:
```diff
@@ -249,6 +250,52 @@ def read_raw_bids(bids_fname, bids_root, verbose=True):
+        The filename corresponding to the empty room.
+        Returns None if no file found.
+    """
+    from mne_bids.write import make_bids_folders
```
why relative import?

jasmainak replied:
because this file is imported in write.py -- I thought it could lead to circular import issue.

sappelhoff replied:
> I thought it could lead to circular import issue.

Have you tried it? Or is it just a suspicion?

agramfort replied:
if we end up with circular imports with such as small project it's a design pb we should fix early

jasmainak replied:
Let's fix this in another issue. I would like to keep the scope of this PR constrained. It already depends on many other PRs in MNE-Python. Opened an issue to deal with it. See: https://github.com/mne-tools/mne-bids/issues/299

jasmainak replied:
okay fair enough. I fixed this in the last commit

agramfort commented on mne_bids/read.py:
```diff
@@ -249,6 +250,52 @@ def read_raw_bids(bids_fname, bids_root, verbose=True):
+        dt = datetime.strptime(params['ses'], '%Y%m%d')
+        delta_t = dt - ref_date
+        if delta_t.total_seconds() < min_seconds:
+            best_er_fname = er_fname
```
`min_seconds = delta_t.total_seconds()` ?

jasmainak replied:
good catch!

agramfort commented on mne_bids/tests/test_read.py:
```diff
@@ -125,3 +128,35 @@ def test_handle_events_reading():
+    er_raw_fname = op.join(data_path, 'MEG', 'sample', 'ernoise_raw.fif')
+    er_raw = mne.io.read_raw_fif(er_raw_fname)
+    er_date = datetime.fromtimestamp(
+        er_raw.info['meas_date'][0]).strftime('%Y%m%d')
```
this is tricky to get write especially if you anonymize. Public API should simplify this.

jasmainak replied:
Let's make an example. I think it will clarify what are the pain points.

jasmainak commented:
@alexrockhill do you want to review this PR? :)

alexrockhill commented:
> @alexrockhill do you want to review this PR? :)

Sure :)

jasmainak commented on examples/convert_empty_room.py:
```diff
@@ -0,0 +1,91 @@
+
+for date in dates:
+	er_bids_basename = 'sub-emptyroom_ses-{0}_task-noise'.format(date)
+	er_raw.info['meas_date'] = (datetime.strptime(date, '%Y%m%d').timestamp(), 0)
```
This is broken at the moment @agramfort 

I think public API from MNE will fix this

sappelhoff commented on examples/convert_empty_room.py:
```diff
@@ -0,0 +1,91 @@
+======================================
+
+This example demonstrates how to store empty room data in BIDS format
+and how to retrieve them.
```
A brief sentence about what emptyroom data is would be nice for EEG only people (who would be confused about this)... Like ".. used for MEG ... see this link"

jasmainak replied:
Okay improved the narrative

sappelhoff commented on examples/convert_empty_room.py:
```diff
@@ -0,0 +1,91 @@
+
+data_path = sample.data_path()
+output_path = op.join(data_path, '..', 'MNE-sample-data-bids')
+raw_fname = op.join(data_path, 'MEG', 'sample', 'sample_audvis_raw.fif')
```
```Python
# paths to our raw data
data_path = sample.data_path()
raw_fname = op.join(data_path, 'MEG', 'sample', 'sample_audvis_raw.fif')

# place to but converted data
output_path = op.join(data_path, '..', 'MNE-sample-data-bids')
```

jasmainak replied:
I changed the order of the variable declarations and renamed `output_path` to `bids_path`. Hopefully it's clearer now.

sappelhoff commented on examples/convert_empty_room.py:
```diff
@@ -0,0 +1,91 @@
+
+###############################################################################
+# Just to illustrate, we can save more than one empty room files. Here, they
+# will all contain the same data but this will not be the case in reality.
```
```suggestion
# will all contain the same data but in your original scripts, you would pass different data.
```

sappelhoff commented on mne_bids/read.py:
```diff
@@ -256,6 +257,53 @@ def read_raw_bids(bids_fname, bids_root, extra_params=None,
+                                    session='**', make_dir=False)
+    search_path = op.join(search_path, '**', '**%s' % ext)
+    er_fnames = glob.glob(search_path)
+
```
Can you re-use https://github.com/mne-tools/mne-bids/blob/d045ec254a0536af8b18976450bf7a12d1c8fd9e/mne_bids/utils.py#L454 somehow?

agramfort commented on examples/convert_empty_room.py:
```diff
@@ -0,0 +1,91 @@
+###############################################################################
+# We also need to specify that the subject ID is 'emptyroom', and that the 
+# task is 'noise'.
+er_bids_basename = 'sub-emptyroom_ses-{0}_task-noise'.format(er_date)
```
make_bids_basename should be able to do this automatically when task=noise and subject=empyroom

so you would not have to do the kung fu with dates above.

jasmainak replied:
are you proposing that `make_bids_basename` take `raw` as an input?

jasmainak replied:
I think the kung fu is much less now that MNE uses `datetime` objects.

agramfort replied:
it feels wrong in terms of design. make_bids_basename should take care of the formatting of bids
filenames. Here we do  the job of this function. How about you use:

make_bids_basename here but enforce that session is a well formatted date if task is called noise?

agramfort commented on examples/convert_empty_room.py:
```diff
@@ -0,0 +1,91 @@
+from mne_bids import get_matched_empty_room
+
+bids_fname = bids_basename + '_meg.fif'
+best_er_fname = get_matched_empty_room(bids_fname, output_path)
```
it would be simpler if get_matched_empty_room could take bids_basename as input. It would abstract away the file type

jasmainak replied:
you wouldn't know whether to look for `meg` or `eeg` etc. if you just gave `bids_basename`. It's fine for writing but not so good for reading files.

alexrockhill commented on mne_bids/read.py:
```diff
@@ -256,6 +257,53 @@ def read_raw_bids(bids_fname, bids_root, extra_params=None,
+
+    if len(er_fnames) == 0:
+        return None
+
```
Maybe more pythonic to define `best_er_fname=None` before the loop. If there's nothing to loop, will just return None.

alexrockhill commented on mne_bids/tests/test_read.py:
```diff
@@ -155,3 +158,35 @@ def test_get_head_mri_trans_ctf():
+
+    er_raw_fname = op.join(data_path, 'MEG', 'sample', 'ernoise_raw.fif')
+    er_raw = mne.io.read_raw_fif(er_raw_fname)
+    er_date = datetime.fromtimestamp(
```
Again fail on windows with anonymize

jasmainak commented on examples/convert_empty_room.py:
```diff
@@ -0,0 +1,91 @@
+###############################################################################
+# The measurement date is
+raw_date = datetime.fromtimestamp(
+    raw.info['meas_date'][0]).strftime('%Y%m%d')
```
I would say that this will be resolved once MNE starts storing the `meas_date` as datetime objects. It will simplify this code.

jasmainak commented on mne_bids/read.py:
```diff
@@ -256,6 +257,53 @@ def read_raw_bids(bids_fname, bids_root, extra_params=None,
+    ref_date = datetime.fromtimestamp(raw.info['meas_date'][0])
+    search_path = make_bids_folders(output_path=bids_root, subject='emptyroom',
+                                    session='**', make_dir=False)
+    search_path = op.join(search_path, '**', '**%s' % ext)
```
Yes I would say you want to match by extension. They are different systems so their noise structure will be different.

jasmainak commented on examples/convert_empty_room.py:
```diff
@@ -0,0 +1,91 @@
+
+# For empty room data we need to specify the recording date in the format
+# YYYYMMDD for the session id.
+er_date = datetime.fromtimestamp(
```
I would rather use a more general Python function where possible. In this case, we know that this empty room will not have timestamp < 0 so we don't need to resort to fancier methods that will confuse the user.

jasmainak commented:
in India, we believe that the good and bad deeds on 1st January get repeated the rest of the year. In that spirit, I made this PR ready to merge ;-) Hopefully CIs are as happy as I am.

sappelhoff commented:
> in India, we believe that the good and bad deeds on 1st January get repeated the rest of the year. In that spirit, I made this PR ready to merge ;-)

Haha, happy new year! This was mostly between @SophieHerbst @agramfort and you, so I'll leave the review to them.

jasmainak commented on examples/convert_mne_sample.py:
```diff
@@ -66,28 +65,6 @@
-if not isinstance(meas_date, datetime):
-    meas_date = datetime.fromtimestamp([0])  # for MNE < v0.20
-er_date = meas_date.strftime('%Y%m%d')
-er_bids_basename = 'sub-emptyroom_ses-{0}_task-noise'.format(er_date)
```
no the MNE-BIDS code will still work with lower versions of MNE. I just wanted to avoid unnecessary complexity in the example.

jasmainak replied:
also, this has not been removed, it is moved to a _new_ example

jasmainak commented:
>  How about you use: make_bids_basename here but enforce that session is a well formatted date if task is called noise?

done now. I think I've addressed all your comments now

jasmainak commented on README.rst:
```diff
@@ -210,38 +210,8 @@ Please see our `contributing guide <https://github.com/mne-tools/mne-bids/blob/m
-   David, O., … Hermes, D. (2019). iEEG-BIDS, extending the Brain Imaging Data
-   Structure specification to human intracranial electrophysiology. Scientific
-   Data, 6, 102. https://doi.org/10.1038/s41597-019-0105-7
+    Appelhoff, S., Sanderson, M., Brooks, T., Vliet, M., Quentin, R., Holdgraf, C., Chaumon, M., Mikulan, E., Tavabi, K., Höchenberger, R., Welke, D., Brunner, C., Rockhill, A., Larson, E., Gramfort, A. and Jas, M. (2019). MNE-BIDS: Organizing electrophysiological data into the BIDS format and facilitating their analysis. Journal of Open Source Software 4: (1896).
```
@sappelhoff @agramfort I updated this. Okay for you?

agramfort replied:
I would ask to cite this one first and then "one of the reference papers".  So have both citations requested.

agramfort commented on examples/convert_empty_room.py:
```diff
@@ -0,0 +1,99 @@
+dates = ['20021204', '20021201', '20021001']
+
+for date in dates:
+    er_bids_basename = 'sub-emptyroom_ses-{0}_task-noise'.format(date)
```
use make_bids_basename here too.

agramfort commented on mne_bids/__init__.py:
```diff
@@ -3,10 +3,10 @@
-from mne_bids.read import read_raw_bids, get_head_mri_trans  # noqa: F401
+from mne_bids.read import read_raw_bids, get_head_mri_trans, get_matched_empty_room  # noqa: F401
+from mne_bids.utils import make_bids_folders  # noqa: F401
 from mne_bids import commands  # noqa: F401
```
all these should be relative imports

jasmainak replied:
this was discussed in https://github.com/mne-tools/mne-bids/pull/248 where we switched to absolute imports because of pep8

jasmainak commented:
comments addressed

agramfort commented on README.rst:
```diff
@@ -210,20 +210,24 @@ Please see our `contributing guide <https://github.com/mne-tools/mne-bids/blob/m
+  Chaumon, M., Mikulan, E., Tavabi, K., Höchenberger, R., Welke, D., Brunner, C.,
+  Rockhill, A., Larson, E., Gramfort, A. and Jas, M. (2019). MNE-BIDS: Organizing
+  electrophysiological data into the BIDS format and facilitating their analysis.
+  Journal of Open Source Software 4: (1896).
```
It should be meg paper here

jasmainak replied:
brrr really sorry about that. Fixed it now.
